### PR TITLE
feat: User defined git repo for GitOps deployment

### PR DESCRIPTION
**Pull request by Ash-exp**
# Description
In Devtron, currently we create a repo for every application for storing GitOps data, if the org doesn’t want to give us create repo permission, it becomes an issue.

In this PR we gave this flexibility to users to choose whether they want us to create the repo or they want to create and provide us the URL

Fixes https://github.com/devtron-labs/devtron/issues/4308

## How Has This Been Tested?
- [x] GItOps Config CRUD
- [x] Helm Chart GitOps Installation
- [x] Helm Chart Chart Group Installation
- [x] Helm Chart Bulk Installation
- [x] Devtron App Level GitOps Config Create and Validate
- [x] CD with GitOps
- [x] CD with Helm
- [x] GitOps Repo Migration

## Checklist:

* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] Does this PR requires documentation updates?
* [ ] I've updated documentation as required by this PR.
* [x] I have performed a self-review of my own code.
* [x] I have commented my code, particularly in hard-to-understand areas.
* [ ] I have tested it for all user roles.
* [ ] I have added all the required unit/api test cases.

## Does this PR introduce a user-facing change?

```release-note

```

### CODE REVIEW

File: pkg/pipeline/DeploymentPipelineConfigService.go [hunk 1]
```diff
@@ -386,24 +399,32 @@ func (impl *CdPipelineConfigServiceImpl) CreateCdPipelines(pipelineCreateRequest
...
-
-		gitopsRepoName, chartGitAttr, err := impl.appService.CreateGitopsRepo(app, pipelineCreateRequest.UserId)
-		if err != nil {
-			impl.logger.Errorw("error in creating git repo", "err", err)
-			return nil, err
-		}
-
-		err = impl.RegisterInACD(gitopsRepoName, chartGitAttr, pipelineCreateRequest.UserId, ctx)
+	if isGitOpsConfigured && isGitOpsRequiredForCD && !pipelineCreateRequest.IsCloneAppReq {
+		chart, err := impl.chartRepository.FindLatestChartForAppByAppId(app.Id)
```
Comment (iamayushm) [inline pkg/pipeline/DeploymentPipelineConfigService.go:403]: use chartService function for this

File: pkg/pipeline/DeploymentPipelineConfigService.go [hunk 2]
```diff
@@ -386,24 +399,32 @@ func (impl *CdPipelineConfigServiceImpl) CreateCdPipelines(pipelineCreateRequest
...
-			impl.logger.Errorw("error in registering app in acd", "err", err)
+			impl.logger.Errorw("Error in fetching latest chart for pipeline", "err", err, "appId", app.Id)
 			return nil, err
 		}
-
-		err = impl.updateGitRepoUrlInCharts(pipelineCreateRequest.AppId, chartGitAttr, pipelineCreateRequest.UserId)
-		if err != nil {
-			impl.logger.Errorw("error in updating git repo url in charts", "err", err)
-			return nil, err
+		if ChartsUtil.IsGitOpsRepoNotConfigured(chart.GitRepoUrl) {
```
Comment (iamayushm) [inline pkg/pipeline/DeploymentPipelineConfigService.go:408]: after using chartService function, implement this method on result struct

File: pkg/pipeline/DeploymentPipelineConfigService.go [hunk 3]
```diff
@@ -386,24 +399,32 @@ func (impl *CdPipelineConfigServiceImpl) CreateCdPipelines(pipelineCreateRequest
...
+			if gitOpsConfig.AllowCustomRepository || chart.IsCustomGitRepository {
```
Comment (iamayushm) [inline pkg/pipeline/DeploymentPipelineConfigService.go:409]: can we move this logic with above check ( if ChartsUtil.IsGitOpsRepoNotConfigured(chart.GitRepoUrl) ) and pass gitOpsConfig.AllowCustomRepository as function parameter

**Review by iamayushm — CHANGES_REQUESTED** [review]

File: pkg/chart/ChartService.go [hunk 4]
```diff
@@ -1720,3 +1750,135 @@ func (impl ChartServiceImpl) GetCustomChartInBytes(chartRefId int) ([]byte, erro
...
 	return manifestByteArr, nil
 }
+
+func (impl ChartServiceImpl) UpdateGitRepoUrlInCharts(appId int, chartGitAttribute *util.ChartGitAttribute, userId int32) error {
+	charts, err := impl.chartRepository.FindActiveChartsByAppId(appId)
+	if err != nil && pg.ErrNoRows != err {
+		return err
+	}
+	for _, ch := range charts {
+		if ChartsUtil.IsGitOpsRepoNotConfigured(ch.GitRepoUrl) {
```
Comment (iamayushm) [inline pkg/chart/ChartService.go:1760]: what if allow custom repository is false in global gitops, we need to check it here also?

Comment (Ash-exp) [reply pkg/chart/ChartService.go:1760]: It has been verified prior to the function call

File: pkg/pipeline/DeploymentPipelineConfigService.go (hunk 2 above)
Comment (Ash-exp) [reply pkg/pipeline/DeploymentPipelineConfigService.go:408]: ChartsUtil.IsGitOpsRepoNotConfigured(chart.GitRepoUrl): is also used at places where we bind the repository struct with JOIN query. There we can not use the above implementation

File: pkg/pipeline/DeploymentPipelineConfigService.go (hunk 1 above)
Comment (Ash-exp) [reply pkg/pipeline/DeploymentPipelineConfigService.go:403]: updated 👍

File: pkg/pipeline/DeploymentPipelineConfigService.go (hunk 3 above)
Comment (Ash-exp) [reply pkg/pipeline/DeploymentPipelineConfigService.go:409]: we cannot combine this condition as we have different business logic based on `if gitOpsConfig.AllowCustomRepository || chart.IsCustomGitRepository`

File: pkg/appStore/deployment/service/AppStoreDeploymentService.go [hunk 5]
```diff
@@ -243,9 +256,46 @@ func (impl AppStoreDeploymentServiceImpl) AppStoreDeployOperationDB(installAppVe
...
+				PerformDefaultValidation: installAppVersionRequest.GitOpsRepoURL == bean2.GIT_REPO_DEFAULT,
+			}
+
+			detailedErrorGitOpsConfigResponse := impl.appStoreDeploymentArgoCdService.ValidateCustomGitRepoURL(validateCustomGitRepoURLRequest)
+			installAppVersionRequest.DetailedErrorGitOpsConfigResponse = &detailedErrorGitOpsConfigResponse
+			if len(detailedErrorGitOpsConfigResponse.StageErrorMap) != 0 {
+				// Found validation err
+				return installAppVersionRequest, nil
+			}
+			installAppVersionRequest.GitOpsRepoURL = installAppVersionRequest.GitOpsRepoURL
```
Comment (iamayushm) [inline pkg/appStore/deployment/service/AppStoreDeploymentService.go:283]: remove this statement

Comment (Ash-exp) [reply pkg/appStore/deployment/service/AppStoreDeploymentService.go]: removed 👍

**Review by nishant-d — APPROVED** [review]

**Pull request merged**